hilti/runtime/include/types/string.h

```diff
@@ -12,27 +12,42 @@ namespace hilti::rt {
 
 namespace string {
 
-/** Computes the length of a UTF8 string.
+/* When processing UTF8, how to handle invalid data not representing UTF8 codepoints. */
+enum class DecodeErrorStrategy : int64_t {
+    IGNORE,  // skip data
+    REPLACE, // replace with a place-holder
+    STRICT   // throw a runtime error
+};
+
+/**
+ * Computes the length of a UTF8 string.
```

Comment:
Let's be very explicit how replacement characters factor into the length. If I understand the implementation correctly we do not return an UTF8 string length, but the number of code points (e.g., we do not handle combining characters). Each inserted replacement characters should contribute one additional code point.